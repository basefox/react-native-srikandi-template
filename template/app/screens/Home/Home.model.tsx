export enum Menu {
  MY_DAY = 'My Day',
  IMPORTANT = 'Important',
  PLANNED = 'Planned',
  TASKS = 'Tasks',
}
