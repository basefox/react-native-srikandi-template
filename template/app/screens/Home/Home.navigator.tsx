import {useNavigation} from '@react-navigation/native';
import {StackNavigationProp} from '@react-navigation/stack';
import {RootStackParamList} from '../../routes/ParamList';

export const HomeNavigator = () => {
  const navigation = useNavigation<StackNavigationProp<RootStackParamList>>();

  function goToTasks(categoryId?: string) {
    navigation.navigate('Tasks', {categoryId});
  }
  return {
    goToTasks,
  };
};
