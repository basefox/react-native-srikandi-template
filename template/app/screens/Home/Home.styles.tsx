import {StyleSheet} from 'react-native';
import {colorBackground, typography} from '@styles/index';

export default StyleSheet.create({
  container: {
    flex: 1,
    padding: 16,
    backgroundColor: colorBackground.white,
  },
  containerBtn: {
    justifyContent: 'flex-end',
    alignSelf: 'flex-end',
  },
  sectionDescription: {
    marginTop: 8,
    fontSize: 18,
    fontWeight: '400',
  },
  highlight: {
    fontWeight: '700',
  },
  subTitle: {
    ...typography.subtitle,
    marginTop: 24,
    marginBottom: 16,
  },
  line: {
    height: 1,
    width: '100%',
    backgroundColor: colorBackground.grey,
    marginVertical: 16,
  },
});
