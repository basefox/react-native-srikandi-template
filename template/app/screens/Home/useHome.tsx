import cuid from 'cuid';
import {useState} from 'react';
import {useDispatch, useSelector} from 'react-redux';
import {selectCategories} from '@appredux/categories/Categories.selector';
import {HomeNavigator} from './Home.navigator';
import {TasksUseCase} from '@domain/TasksUseCase';

useHome.dependencies = {
  navigator: HomeNavigator,
  tasksUseCase: TasksUseCase,
};

export function useHome() {
  const {navigator, tasksUseCase} = useHome.dependencies;
  const nav = navigator();

  const dispatch = useDispatch();
  const categories = useSelector(selectCategories);
  const [showFormCreateList, setShowCreateList] = useState<boolean>(false);

  function onAddCategory(title: string) {
    setShowCreateList(false);
    tasksUseCase.addCategoryTask({
      id: cuid(),
      title: title,
      tasks: [],
    })(dispatch);
  }

  function onPressCategoryItem({id}: {id: string}) {
    nav.goToTasks(id);
  }

  return {
    data: {
      categories,
    },
    state: {
      showFormCreateList,
    },
    actions: {
      setShowCreateList,
      onAddCategory,
      onPressCategoryItem,
    },
  };
}
