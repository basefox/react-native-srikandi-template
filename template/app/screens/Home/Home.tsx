/**
 * Sample React Native App
 * https://github.com/facebook/react-native
 *
 * Generated with the TypeScript template
 * https://github.com/react-native-community/react-native-template-typescript
 *
 * @format
 */

import React from 'react';
import {ScrollView, StatusBar, useColorScheme, View} from 'react-native';
import {BaseScreen, Button, Text} from '@components/index';
import {Icon} from '@components/TabIcon';
import {IcSun, IcStar, IcTask, IcCalendar, IcPlus} from '@assets/svgs';
import {colorBackground, colorBase} from '@styles/index';
import {ListMenu, ListCategory} from './contents/ListMenu';
import {FormCreateTask} from './contents/FormCreateTask';
import {Menu} from './Home.model';
import {useHome} from './useHome';
import styles from './Home.styles';

const Line = () => <View style={styles.line} />;

Home.dependencies = {
  useActions: useHome,
};

function Home() {
  const {useActions} = Home.dependencies;
  const isDarkMode = useColorScheme() === 'dark';
  const {state, actions, data} = useActions();
  const {showFormCreateList} = state;
  const {setShowCreateList, onAddCategory, onPressCategoryItem} = actions;
  const {categories} = data;

  return (
    <BaseScreen style={styles.container}>
      <StatusBar barStyle={isDarkMode ? 'light-content' : 'dark-content'} />
      <ScrollView contentInsetAdjustmentBehavior="automatic">
        <ListMenu
          data={[
            {
              title: Menu.MY_DAY,
              leftIcon: (
                <Icon Icon={IcSun} size={18} color={colorBackground.grey} />
              ),
              count: 0,
            },
            {
              title: Menu.IMPORTANT,
              leftIcon: (
                <Icon Icon={IcStar} size={18} color={colorBackground.grey} />
              ),
              count: 0,
            },
            {
              title: Menu.PLANNED,
              leftIcon: (
                <Icon
                  Icon={IcCalendar}
                  size={18}
                  color={colorBackground.grey}
                />
              ),
              count: 0,
            },
            {
              title: Menu.TASKS,
              leftIcon: (
                <Icon Icon={IcTask} size={18} color={colorBackground.grey} />
              ),
              count: 0,
            },
          ]}
          onPress={() => {}}
        />
        <Line />
        <Button
          variant="nude"
          icon={IcPlus}
          label="Add List"
          onPress={() => {
            setShowCreateList(true);
          }}
          containerStyle={styles.containerBtn}
        />
        <ListCategory
          data={categories.map(category => {
            return {
              id: category.id,
              title: category.title,
              count: category.tasks.length,
            };
          })}
          onPress={category => onPressCategoryItem({id: category.id})}
        />
      </ScrollView>
      {showFormCreateList && <FormCreateTask onSave={onAddCategory} />}
    </BaseScreen>
  );
}

export default Home;
