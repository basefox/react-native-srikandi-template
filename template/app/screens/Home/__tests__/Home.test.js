import React from 'react';
import {render} from '@testing-library/react-native';
import {useActionMock} from '@utils/tests/useActionMock';
import Home from '../Home';

jest.mock('@components/index', () => ({
  BaseScreen: 'BaseScreen',
  Button: 'Button',
}));
jest.mock('@components/TabIcon', () => 'Icon');
jest.mock('@assets/svgs', () => ({
  IcSun: 'IcSun',
  IcStar: 'IcStar',
  IcTask: 'IcTask',
  IcCalendar: 'IcCalendar',
  IcPlus: 'IcPlus',
}));

describe('Home', () => {
  const useHomeMock = {
    data: {
      categories: [],
    },
    state: {
      showFormCreateList: false,
    },
    actions: {
      setShowCreateList: jest.fn(),
      onAddCategory: jest.fn(),
      onPressCategoryItem: jest.fn(),
    },
  };

  Home.dependencies = {
    useActions: useActionMock(useHomeMock),
  };

  test('should render correctly', () => {
    const {toJSON} = render(<Home />);
    expect(toJSON()).toMatchSnapshot();
  });
});
