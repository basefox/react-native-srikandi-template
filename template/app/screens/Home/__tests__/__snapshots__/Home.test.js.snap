// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should render correctly 1`] = `
<BaseScreen
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "padding": 16,
    }
  }
>
  <RCTScrollView
    contentInsetAdjustmentBehavior="automatic"
  >
    <View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "opacity": 1,
            "paddingVertical": 8,
          }
        }
      >
        <View
          style={
            Object {
              "marginRight": 8,
            }
          }
        >
          <Icon
            Icon="IcSun"
            color="#bbc2d8"
            size={18}
          />
        </View>
        <Text
          style={
            Object {
              "color": "#2E434D",
              "flex": 1,
              "fontFamily": "ubuntu-medium",
              "fontSize": 16,
              "lineHeight": 21,
            }
          }
        >
          My Day
        </Text>
      </View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "opacity": 1,
            "paddingVertical": 8,
          }
        }
      >
        <View
          style={
            Object {
              "marginRight": 8,
            }
          }
        >
          <Icon
            Icon="IcStar"
            color="#bbc2d8"
            size={18}
          />
        </View>
        <Text
          style={
            Object {
              "color": "#2E434D",
              "flex": 1,
              "fontFamily": "ubuntu-medium",
              "fontSize": 16,
              "lineHeight": 21,
            }
          }
        >
          Important
        </Text>
      </View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "opacity": 1,
            "paddingVertical": 8,
          }
        }
      >
        <View
          style={
            Object {
              "marginRight": 8,
            }
          }
        >
          <Icon
            Icon="IcCalendar"
            color="#bbc2d8"
            size={18}
          />
        </View>
        <Text
          style={
            Object {
              "color": "#2E434D",
              "flex": 1,
              "fontFamily": "ubuntu-medium",
              "fontSize": 16,
              "lineHeight": 21,
            }
          }
        >
          Planned
        </Text>
      </View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "opacity": 1,
            "paddingVertical": 8,
          }
        }
      >
        <View
          style={
            Object {
              "marginRight": 8,
            }
          }
        >
          <Icon
            Icon="IcTask"
            color="#bbc2d8"
            size={18}
          />
        </View>
        <Text
          style={
            Object {
              "color": "#2E434D",
              "flex": 1,
              "fontFamily": "ubuntu-medium",
              "fontSize": 16,
              "lineHeight": 21,
            }
          }
        >
          Tasks
        </Text>
      </View>
      <View
        style={
          Object {
            "backgroundColor": "#bbc2d8",
            "height": 1,
            "marginVertical": 16,
            "width": "100%",
          }
        }
      />
      <NudeButton
        containerStyle={
          Object {
            "alignSelf": "flex-end",
            "justifyContent": "flex-end",
          }
        }
        icon={
          <Icon
            Icon="IcPlus"
            color="#B31E72"
            size={18}
          />
        }
        onPress={[Function]}
        title="Add List"
      />
    </View>
  </RCTScrollView>
</BaseScreen>
`;
